ng-dev/pr/merge/merge-pull-request.ts

```diff
@@ -36,17 +37,18 @@ export async function mergePullRequest(prNumber: number, flags: PullRequestMerge
 
   const tool = await createPullRequestMergeTool(flags);
 
-  // Perform the merge. Force mode can be activated through a command line flag.
-  // Alternatively, if the merge fails with non-fatal failures, the script
-  // will prompt whether it should rerun in force mode.
-  if (!(await performMerge(false))) {
+  // Perform the merge. If the merge fails with non-fatal failures, the script
+  // will prompt whether it should rerun in force mode with the ignored failure.
+  if (!(await performMerge())) {
     process.exit(1);
   }
 
   /** Performs the merge and returns whether it was successful or not. */
-  async function performMerge(ignoreFatalErrors: boolean): Promise<boolean> {
+  async function performMerge(
+    validationConfig: PullRequestValidationConfig = new PullRequestValidationConfig(),
+  ): Promise<boolean> {
     try {
-      await tool.merge(prNumber, /* force */ ignoreFatalErrors);
+      await tool.merge(prNumber, validationConfig);
```

Comment:
Rather than recursively calling on an overall failure, would it make more sense to have the force prompt occur from within the validation?

In the `PullRequestValidation` abstract class you could expand it to:

```ts
export abstract class PullRequestValidation {
  abstract canBeForced: boolean;

  constructor(
    protected name: keyof PullRequestValidationConfig,
    protected _createError: PullRequestValidationErrorCreateFn,
  ) {}

  abstract internalAssert(pullRequest: PullRequestFromGithub): Promise<void>;

  async assert(pullRequest: PullRequestFromGithub) {
    try {
      this.internalAssert(pullRequest);
    } catch (e) {
      if (this.canBeForced && await prompt('Continue anyway?')) {
        return;
      }
      throw e;
    }
  }
}

```

Then the forcing logic becomes localize to the specific validation, and we can continue on directly from there.

Reply:
Done. I love it! that made this much more simple and even more readable.

One note: Instead of making the internal assert method typed in the base class, I'm leaving that to the consumer because every validation has different parameters and typing this with a generic (and combining the args with the validation config) is not worth the effort.